biobalm/_sd_attractors/attractor_candidates.py

```diff
@@ -0,0 +1,760 @@
+"""
+Provides functions that are used to identify and reduce the set of attractor candidate states.
+
+Note that most of these functions assume they are operating on the network or petri net
+percolated to the subspace of the relevant node.
+"""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Literal, cast
+
+if TYPE_CHECKING:
+    from biobalm.succession_diagram import SuccessionDiagram
+    from biobalm.types import BooleanSpace
+    from networkx import DiGraph  # type: ignore
+
+import random
+import biobalm
+from biodivine_aeon import Bdd, AsynchronousGraph
+from biobalm.trappist_core import compute_fixed_point_reduced_STG
+from biobalm.symbolic_utils import state_list_to_bdd, valuation_to_state, state_to_bdd
+
+try:
+    PINT_AVAILABLE = True
+    import biobalm._pint_reachability as pint
+except ModuleNotFoundError:
+    PINT_AVAILABLE = False
+
+
+CANDIDATES_HARD_LIMIT = 100_000
+"""
+If there are more than this amount of attractor candidates, then
+attractor detection will fail with a runtime error. This is mainly to
+avoid out-of-memory errors or crashing clingo.
+"""
+
+RETAINED_SET_OPTIMIZATION_THRESHOLD = 100
+"""
+If there are more than this amount of attractor candidates, we will
+try to optimize the retained set using ASP (if enabled).
+"""
+
+
+def compute_attractor_candidates(
+    sd: SuccessionDiagram,
+    node_id: int,
+    greedy_asp_minification: bool,
+    simulation_minification: bool,
+    pint_minification: bool,
+) -> list[BooleanSpace]:
+    """
+    Compute an optimized list of candidate states that is guaranteed
+    to cover every attractor in the specified SD node.
+
+    It should hold that for every attractor in the node's sub-space outside
+    of the child sub-spaces (if known), the list contains at least one state
+    from this attractor. However, it is not guaranteed that each candidate
+    covers some attractor, so a non-empty list of candidates can
+    still correspond to an empty list of attractors.
+
+    The method starts by computing an NFVS of the network percolated to the
+    node's space, and then assigns each NFVS node a value towards which
+    it tends to. This transformation satisfies that the fixed-points in the
+    resulting network cover the attractors in the original sub-space.
+
+    Subsequently, the method can further optimize this set of candidates.
+
+    ### Greedy ASP minification
```

Comment:
Not sure if the autodoc is going to like these headings. Did you test it? We can always fix it later if needed, so it's not a big deal.

Reply:
Good point, I'll check it.